src/main/java/org/folio/circulation/resources/ClaimItemReturnedResource.java

```diff
@@ -0,0 +1,74 @@
+import static org.folio.circulation.support.ValidationErrorFailure.singleValidationError;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.folio.circulation.StoreLoanAndItem;
+import org.folio.circulation.domain.ClaimItemReturnedRequest;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.validation.LoanValidator;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.ItemRepository;
+import org.folio.circulation.support.NoContentResult;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.RouteRegistration;
+import org.folio.circulation.support.http.server.WebContext;
+
+import io.vertx.core.http.HttpClient;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+public class ClaimItemReturnedResource extends Resource {
+  public ClaimItemReturnedResource(HttpClient client) {
+    super(client);
+  }
+
+  @Override
+  public void register(Router router) {
+    new RouteRegistration("/circulation/loans/:id/claim-item-returned", router)
+      .create(this::claimItemReturned);
+  }
+
+  private void claimItemReturned(RoutingContext routingContext) {
+    validateRequest(routingContext)
+      .after(this::processClaimReturned)
+      .thenApply(NoContentResult::from)
+      .thenAccept(result -> result.writeTo(routingContext.response()));
+  }
+
+  private CompletableFuture<Result<Loan>> processClaimReturned(RoutingContext routingContext) {
+    final WebContext context = new WebContext(routingContext);
+    final Clients clients = Clients.create(context, client);
+
+    final LoanRepository loanRepository = new LoanRepository(clients);
+    final ItemRepository itemRepository = noLocationMaterialTypeAndLoanTypeInstance(clients);
+    final StoreLoanAndItem storeLoanAndItem = new StoreLoanAndItem(loanRepository, itemRepository);
+
+    final ClaimItemReturnedRequest request = ClaimItemReturnedRequest.from(routingContext);
+
+    return succeeded(request)
+      .after(req -> loanRepository.getById(req.getLoanId()))
+      .thenApply(LoanValidator::refuseWhenLoanIsClosed)
+      .thenApply(loan -> makeLoanAndItemClaimedReturned(loan, request))
+      .thenCompose(r -> r.after(storeLoanAndItem::updateLoanAndItemInStorage));
+  }
+
+  private Result<Loan> makeLoanAndItemClaimedReturned(
+    Result<Loan> loanResult, ClaimItemReturnedRequest request) {
+
+    return loanResult.map(loan -> loan
+      .claimItemReturned(request.getComment(), request.getItemClaimedReturnedDateTime()));
+  }
+
+  private Result<RoutingContext> validateRequest(RoutingContext routingContext) {
+    return succeeded(routingContext).failWhen(
+      ctx -> succeeded(ctx.getBodyAsJson().getString(ITEM_CLAIMED_RETURNED_DATE) == null),
+      request -> singleValidationError("Item claimed returned date is a required field",
+        ITEM_CLAIMED_RETURNED_DATE, null));
+  }
```

Comment:
Could this be moved to within the `ClaimItemReturnedRequest` class during the construction of the instance?

Reply:
I don't think it is a good idea. It couples the POJO class with validation.

Reply:
> I don't think it is a good idea. It couples the POJO class with validation.

Ok. 

To me, the `ClaimItemReturnedRequest` represents the incoming request to claim that an item is returned. I think only creating this when the request is valid encapsulates rules like required properties.

Are you suggesting that should only be the deserialization of that from the HTTP request and be ignorant of whether it is valid.

And then any uses of that need to be aware that it could be invalid?

Reply:
>  I think only creating this when the request is valid encapsulates rules like required properties.
Right, but there might be different sources of the properties for example HTTP, messaging, file, etc. To me it is a controller responsibility to parse the properties and make sure they are valid. POJO is just a class to transfer the data it does not assure data consistency.  

> Are you suggesting that should only be the deserialization of that from the HTTP request and be ignorant of whether it is valid.
Hmm... Will move deserialization on controller level.

Reply:
> > I think only creating this when the request is valid encapsulates rules like required properties.
> 
> Right, but there might be different sources of the properties for example HTTP, messaging, file, etc. 

Sure, I can understand the need to separate the initial parsing from validation. If we are going to do that, I suggest we model that as a `request` and a `command` class.

I think it is speculation that we might have multiple sources for an incoming request. FOLIO currently only uses HTTP for imperative requests.

> To me it is a controller responsibility to parse the properties and make sure they are valid. POJO is just a class to transfer the data it does not assure data consistency.

I think that might be true when using an MVC based framework or library.

> > Are you suggesting that should only be the deserialization of that from the HTTP request and be ignorant of whether it is valid.
> 
> Hmm... Will move deserialization on controller level.

What are you referring to as the controller? The resource class?

It appears that recent changes have moved both the deserialization and validation responsibilities out of this class. Is that the case?

Reply:
> It appears that recent changes have moved both the deserialization and validation responsibilities out of this class. Is that the case?

Yes, I've moved it to resource class. 

> What are you referring to as the controller? The resource class?

Yes, by controller I mean resource class.

Reply:
It is clear from this conversation that there needs to be a better described pattern for doing this in mod-circulation.

The `ClaimItemReturnedRequest` is currently defined within the domain package. To me, that suggests its intent is to handle domain logic. If instead, it is only to deserialize the incoming JSON over HTTP, then I think we need to consider moving this to do a different pattern and having a separate class for the domain concept.

Reply:
Do you mean business entities (e.g. request, loan, item, etc.) by domain? Yes, `ClaimItemReturnedRequest` it is more like DTO (data transfer object), it is used to transfer data from client only.

I'd use `dto` name for such classes, but I have not seen this term in FOLIO, maybe `resources/requests(responses - can be added in future)`.

What do you think?

Reply:
> Do you mean business entities (e.g. request, loan, item, etc.) by domain? Yes, `ClaimItemReturnedRequest` it is more like DTO (data transfer object), it is used to transfer data from client only.

I was only referring to the package name. I think if we want to clearly separate these concerns, then I think we need two classes, one for serialization (DTO) and one in the domain.

> I'd use `dto` name for such classes, but I have not seen this term in FOLIO, maybe `resources/requests(responses - can be added in future)`.

I agree, naming is a challenge, and we likely need a convention. I haven't noticed DTO used, the RAML Module Builder's generated classes are effectively DTO for both the API and persistence.

If we are to have two classes, I think we have two broad options:
* Use the same name, in two packages (e.g. `serialization.http` and `domain`)
* Use a different postfix `Request` or `DTO` (serlialization) and `Command` (domain)

I'm wondering if this is too big to consider for this change.

Reply:
> I'm wondering if this is too big to consider for this change.

Agree, it is too big I think. I'm not sure why do we need a command class/class in domain. This `ClaimItemReturnedRequest` is only used within the resource class, it has not been used on other layers of the system.

Will it be fine if I move it to `serialization.http` package only?

Reply:
I think we leave it as is and I'll approve the pull request.

(The command and request distinction is if we wanted to the serialization and validation to be separate and encapsulated, rather than moving it to the resource class)